### PR TITLE
[SYCL][FPGA] Add host compilation check to intel-fpga-local test

### PR DESCRIPTION
**Pull request by vmaksimo**

### CODE REVIEW

File: clang/test/CodeGenSYCL/intel-fpga-local.cpp [hunk 1]
```diff
@@ -1,49 +1,53 @@
...
+//CHECK-DEVICE: [[ANN11:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{pump:2}
+//CHECK-DEVICE: [[ANN12:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{merge:foo:depth}
+//CHECK-DEVICE: [[ANN13:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{merge:bar:width}
+//CHECK-DEVICE: [[ANN14:@.str.[0-9]*]] = {{.*}}{max_replicates:2}
+//CHECK-DEVICE: [[ANN15:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{simple_dual_port:1}
+//CHECK-DEVICE: [[ANN6:@.str.[0-9]*]] = {{.*}}{memory:BLOCK_RAM}{sizeinfo:4}
+//CHECK-DEVICE: [[ANN7:@.str.[0-9]*]] = {{.*}}{memory:MLAB}{sizeinfo:4}
+//CHECK-DEVICE: [[ANN8:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{bankwidth:8}
+//CHECK-DEVICE: [[ANN9:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{max_private_copies:4}
+//CHECK-BOTH: [[ANN16:@.str[.0-9]*]] = {{.*}}foobar
```
Comment (mlychkov) [inline clang/test/CodeGenSYCL/intel-fpga-local.cpp]: Is this label differ on host and device IRs?

Comment (vmaksimo) [reply clang/test/CodeGenSYCL/intel-fpga-local.cpp]: No, because the annotate attribute is not SYCL-specific.

Comment (mlychkov) [reply clang/test/CodeGenSYCL/intel-fpga-local.cpp]: Then maybe to revert change in label's regexp?

Comment (vmaksimo) [reply clang/test/CodeGenSYCL/intel-fpga-local.cpp]: I misunderstood you at first -  it has no difference in the annotation itself, but the numbering differs (the first global string has `@str`, not `@str.[0-9]+`

Comment (mlychkov) [reply clang/test/CodeGenSYCL/intel-fpga-local.cpp]: Ok? that is what I wanted to know.

File: clang/test/CodeGenSYCL/intel-fpga-local.cpp [hunk 2]
```diff
@@ -1,49 +1,53 @@
-// RUN: %clang_cc1 -x c++ -triple spir64-unknown-linux-sycldevice -std=c++11 -disable-llvm-passes -fsycl-is-device -emit-llvm %s -o - | FileCheck %s
+// RUN: %clang_cc1 -x c++ -triple spir64-unknown-linux-sycldevice -std=c++11 -disable-llvm-passes -fsycl-is-device -emit-llvm %s -o - | FileCheck %s -check-prefixes CHECK-DEVICE,CHECK-BOTH
```
Comment (AGindinson) [inline clang/test/CodeGenSYCL/intel-fpga-local.cpp]:
```suggestion
// RUN: %clang_cc1 -x c++ -triple spir64-unknown-unknown-sycldevice -std=c++11 -disable-llvm-passes -fsycl-is-device -emit-llvm %s -o - | FileCheck %s -check-prefixes CHECK-DEVICE,CHECK-BOTH
```

File: clang/test/CodeGenSYCL/intel-fpga-local.cpp [hunk 3]
```diff
@@ -1,49 +1,53 @@
...
-//CHECK: [[ANN7:@.str.[0-9]*]] = {{.*}}{memory:MLAB}{sizeinfo:4}
-//CHECK: [[ANN8:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{bankwidth:8}
-//CHECK: [[ANN9:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{max_private_copies:4}
-//CHECK: [[ANN16:@.str.[0-9]*]] = {{.*}}foobar
-//CHECK: [[ANN17:@.str.[0-9]*]] = {{.*}}{memory:MLAB}{sizeinfo:4,500}
-//CHECK: [[ANN18:@.str.[0-9]*]] = {{.*}}{memory:BLOCK_RAM}{sizeinfo:4,10,2}
-//CHECK: [[ANN19:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{numbanks:4}{bank_bits:4,5}
-//CHECK: [[ANN20:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4,10,2}{bankwidth:16}{numbanks:2}{bank_bits:0}
-//CHECK: [[ANN21:@.str.[0-9]*]] = {{.*}}{memory:MLAB}{sizeinfo:4}{numbanks:8}{bank_bits:5,4,3}
+//CHECK-DEVICE: [[ANN1:@.str]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{numbanks:4}
```
Comment (AGindinson) [inline clang/test/CodeGenSYCL/intel-fpga-local.cpp]: Very minor, just for the sake of unification. Can be applied to the whole test:
```suggestion
// CHECK-DEVICE: [[ANN1:@.str]] = {{.*}}{memory:DEFAULT}{sizeinfo:4}{numbanks:4}
```

File: clang/test/CodeGenSYCL/intel-fpga-local.cpp [hunk 4]
```diff
@@ -1,49 +1,53 @@
...
+//CHECK-DEVICE: [[ANN20:@.str.[0-9]*]] = {{.*}}{memory:DEFAULT}{sizeinfo:4,10,2}{bankwidth:16}{numbanks:2}{bank_bits:0}
+//CHECK-DEVICE: [[ANN21:@.str.[0-9]*]] = {{.*}}{memory:MLAB}{sizeinfo:4}{numbanks:8}{bank_bits:5,4,3}
 
-//CHECK: @llvm.global.annotations
-//CHECK-SAME: { i8* addrspacecast (i8 addrspace(1)* bitcast (i32 addrspace(1)* @_ZZ3quxiE5a_one to i8 addrspace(1)*) to i8*)
-//CHECK-SAME: [[ANN1]]{{.*}}i32 157
-//CHECK-SAME: { i8* addrspacecast (i8 addrspace(1)* bitcast (i32 addrspace(1)* @_ZZ3quxiE5b_two to i8 addrspace(1)*) to i8*)
-//CHECK-SAME: [[ANN16]]{{.*}}i32 161
+//CHECK-BOTH: @llvm.global.annotations
+//CHECK-DEVICE-SAME: { i8* addrspacecast (i8 addrspace(1)* bitcast (i32 addrspace(1)* @_ZZ3quxiE5a_one to i8 addrspace(1)*) to i8*)
```
Comment (AGindinson) [inline clang/test/CodeGenSYCL/intel-fpga-local.cpp]: I used to think that the FileCheck runs for separate check-prefixes do not share the same `CHECK-` lines, and therefore the `CHECK-DEVICE-SAME` wouldn't "attach" to a `CHECK-BOTH`. Am I simply wrong here?

Comment (vmaksimo) [reply clang/test/CodeGenSYCL/intel-fpga-local.cpp]: I can answer this way - the test is correctly passed, so I do an assumption to think that `CHECK-DEVICE-SAME` is really "attached" to `CHECK-BOTH`.

Comment (AGindinson) [reply clang/test/CodeGenSYCL/intel-fpga-local.cpp]: Yes, I've checked this via:
`FileCheck test.txt --input-file test.txt --check-prefixes=CHECK-ALL,CHECK-SPECIFIC
```
// CHECK-ALL: All // CHECK-SPECIFIC-SAME: Specific                                                                                                                                             AllSpecific
```
Cool to know!

**Comment by vmaksimo** [comment]
> Actually, this is another example of a test where we might want to split the testing of separate attributes into multiple kernels. In case you're fine with adding a `TODO` within this PR, please do.

Actually, I don't think that it's worth it in this case. I agree that it's hard to extend test with LLVM IR source, but for C++/SYCL code there are no such problems.

**Comment by AGindinson** [comment]
> > Actually, this is another example of a test where we might want to split the testing of separate attributes into multiple kernels. In case you're fine with adding a `TODO` within this PR, please do.
> 
> Actually, I don't think that it's worth it in this case. I agree that it's hard to extend test with LLVM IR source, but for C++/SYCL code there are no such problems.

While updating a thus-structured LLVM IR-based test is obviously worse, one still can feel the pain when reordering the "self-explaining" variable names in the FileCheck directives. Yes, `VAR_TWO`, `VAR_FIVE`, I'm talking to you.

**Pull request merged**